llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp

```diff
@@ -1284,6 +1284,14 @@ Instruction *InstCombinerImpl::visitLShr(BinaryOperator &I) {
     return NewSub;
   }
 
+  // ((X % 2) + (Y % 2)) / 2 --> (X & Y & 1)
+  if (match(Op0,
+            m_Add(m_And(m_Value(X), m_One()), m_And(m_Value(Y), m_One()))) &&
```

Comment:
Think this can be generalized as just `(X + Y)/2` -> `(X & Y & 1)` where we can prove `X u<= 1` and `Y u<= 1`.

Reply:
No. Please use `computeKnownBits(...).countMaxActiveBits() <= 1`.